hera_sim/simulate.py

```diff
@@ -31,226 +39,386 @@ def new_func(*args, **kwargs):
+        """Simulation times in JD."""
         return np.unique(self.data.time_array)
 
+    @cached_property
+    def pols(self):
+        """Array of polarization strings."""
+        return self.data.get_pols()
+
     def apply_defaults(self, config, refresh=True):
-        # TODO: docstring
         """
+        Apply the provided default configuration.
+
+        Equivalent to calling ``hera_sim.defaults`` with the same parameters.
+        See ``hera_sim.defaults.set`` documentation for further details.
         """
-        # actually apply the default settings
         defaults.set(config, refresh=refresh)
 
-    def add(self, component, **kwargs):
-        # TODO: docstring
+    def calculate_filters(
+        self,
+        *,
+        delay_filter_kwargs: Optional[Dict[str, Union[float, str]]] = None,
+        fringe_filter_kwargs: Optional[Dict[str, Union[float, str, np.ndarray]]] = None,
+    ):
         """
+        Pre-compute fringe-rate and delay filters for the entire array.
+
+        Parameters
+        ----------
+        delay_filter_kwargs
+            Extra parameters necessary for generating a delay filter. See
+            :func:`utils.gen_delay_filter` for details.
+        fringe_filter_kwargs
+            Extra parameters necessary for generating a fringe filter. See
+            :func:`utils.gen_fringe_filter` for details.
         """
-        # find out whether to add and/or return the component
-        add_vis = kwargs.pop("add_vis", True)
-        ret_vis = kwargs.pop("ret_vis", False)
-
-        # find out whether the data application should be filtered
-        vis_filter = kwargs.pop("vis_filter", None)
-
-        # take out the seed kwarg so as not to break initializor
-        seed = kwargs.pop("seed", -1)
+        delay_filter_kwargs = delay_filter_kwargs or {}
+        fringe_filter_kwargs = fringe_filter_kwargs or {}
+        self._calculate_delay_filters(**delay_filter_kwargs)
+        self._calculate_fringe_filters(**fringe_filter_kwargs)
 
-        # get the model for the desired component
-        model, is_class = self._get_component(component)
-
-        # make a new entry in the antpairpol cache
-        self._antpairpol_cache[model] = []
-
-        # get a reference to the cache
-        antpairpol_cache = self._antpairpol_cache[model]
-
-        # make sure to keep the key handy in case it's a class
-        model_key = model
-
-        # instantiate the class if the component is a class
-        if is_class:
+    def add(
+        self,
+        component: Component,
+        *,
+        add_vis: bool = True,
+        ret_vis: bool = False,
+        seed: Optional[Union[str, int]] = None,
+        vis_filter: Optional[Sequence] = None,
+        name: Optional[str] = None,
+        **kwargs,
+    ):
+        """
+        Simulate an effect then apply and/or return the result.
+
+        Parameters
+        ----------
+        component: str or SimulationComponent subclass
```

Comment:
again with the spaces before colons, and use `:class:` for the sim component.

Reply:
Done.